cumulusci/tasks/bulkdata/step.py

```diff
@@ -285,7 +285,9 @@ def load_records(self, records):
 
     def _batch(self, records):
         """Return a generator of generators, where each child generator is batched."""
-        for batch in batch_iterator(records, self.api_options.get("batch_size", 10000)):
+        for batch in batch_iterator(
+            records, self.fields, self.api_options.get("batch_size", 10000)
+        ):
             yield self._csv_generator(batch)
```

Comment:
I think that we're making things harder for ourselves by trying to produce a batch of the correct size first, then passing it along to `_csv_generator`. It means that batch_iterator needs to do a bunch of work to anticipate how big the records will be in CSV format and know details about the CSV dialect that will be used.

Instead I think we could replace `_batch` and `_csv_generator` with a function that does something like this:

1. Iterate over `records`
2. For each record, use a csv writer to serialize it as CSV, and read the serialized value into a variable `row`.
3. Check the length of `row`. If it will not put the current batch over the limit, append the row to a list for the current batch. If it will put the current batch over the limit, yield an iterator over the current batch, then start a new list for the next batch.

The downside is that we have to store a whole batch in memory. It's probably possible to avoid this by being more clever with the iterators but I'm having trouble figuring it out at the moment.

Reply:
I like it, this would simplify things a bit. Let me play around with reworking and see what I can come up with.